staging/kos/pkg/controllers/subnet/validator_test.go

```diff
@@ -0,0 +1,506 @@
+	k8scache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	netv1a1 "k8s.io/examples/staging/kos/pkg/apis/network/v1alpha1"
+	kosfake "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/fake"
+	koscsv1a1 "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/typed/network/v1alpha1"
+	kosinformers "k8s.io/examples/staging/kos/pkg/client/informers/externalversions"
+)
+
+const (
+	// Values used to initialize test subnets.
+	ns1, ns2     = "ns1", "ns2"
+	name1, name2 = "s1", "s2"
+	cidr1, cidr2 = "192.168.10.0/24", "192.168.11.0/24"
+	vni1, vni2   = 1, 2
+	uid1, uid2   = "1", "2"
+	rv           = "1"
+)
+
+// Each subnet validator test goes through a sequence of one or more rounds, and
+// each round has its own expected output.
+type validatorTestRound struct {
+	// The subnets that we expect to exist in the API server at the end of the
+	// test round. This is the expected output.
+	expectedSubnets []netv1a1.Subnet
+
+	// The function to invoke before transitioning to the next round. This
+	// typically deletes/creates a subnet, so that the next round can check that
+	// the validator correctly reconsidered the other subnets.
+	transitionToNextRound func(koscsv1a1.NetworkV1alpha1Interface) error
+}
+
+type validatorTestCase struct {
+	// Subnets that exist in the API server when the test starts.
+	initialSubnets []runtime.Object
+
+	// Each test has one or more rounds, each round has its own expected output.
+	rounds []validatorTestRound
+
+	// reaction can be used to inject failures in the API calls the subnet
+	// validator does.
+	reaction k8stesting.ReactionFunc
+}
+
+func TestSubnetValidator(t *testing.T) {
+	testCases := map[string]validatorTestCase{
+		// Test cases checking that valid is monotonic.
+		"only existing subnet is valid, stays valid": {
+			initialSubnets: []runtime.Object{
+				newSubnet1(valid),
+			},
+			rounds: []validatorTestRound{
+				{
+					expectedSubnets: []netv1a1.Subnet{
+						*newSubnet1(valid),
+					},
+				},
+			},
+		},
+
+		// Test cases checking that valid subnets become valid.
+		"only existing subnet becomes valid": {
+			initialSubnets: []runtime.Object{
+				newSubnet1(),
+			},
+			rounds: []validatorTestRound{
+				{
+					expectedSubnets: []netv1a1.Subnet{
+						*newSubnet1(valid),
+					},
+				},
+			},
+		},
+		"two subnets with different VNIs become valid": {
+			initialSubnets: []runtime.Object{
+				newSubnet1(),
+				newSubnet2(),
+			},
+			rounds: []validatorTestRound{
+				{
+					expectedSubnets: []netv1a1.Subnet{
+						*newSubnet1(valid),
+						*newSubnet2(valid),
+					},
+				},
+			},
+		},
+		"two non-conflicting subnets with same VNI become valid": {
+			initialSubnets: []runtime.Object{
+				newSubnet1(),
+				newSubnet2(ns(ns1), vni(vni1)),
+			},
+			rounds: []validatorTestRound{
+				{
+					expectedSubnets: []netv1a1.Subnet{
+						*newSubnet1(valid),
+						*newSubnet2(ns(ns1), vni(vni1), valid),
+					},
+				},
+			},
+		},
+
```

Comment:
How about a test case for the same VNI being used in two namespaces?

Reply:
Yes, I plan to add more tests than those currently there, which act as an initial core to drive discussion on the general outline of unit tests for the subnet validator (and to a lesser extent for the other kos controllers).

Reply:
I just noticed though that in [this](https://github.com/MikeSpreitzer/kube-examples/pull/81#discussion_r337855636) comment we discussed removing the constraint that a VNI must be confined to a single K8s API namespace. If indeed that's what we plan to do, we might skip adding unit tests that make sure that constraint is enforced.